packages/react-tooltip/src/components/Tooltip/useTooltip.tsx

```diff
@@ -30,53 +38,214 @@ export const useTooltip = (
   ref: React.Ref<HTMLElement>,
   defaultProps?: TooltipProps,
 ): TooltipState => {
-  const [target, setTarget] = React.useState<HTMLElement | null>(null);
+  const [visible, setVisible] = React.useState(false);
+  const context = React.useContext(TooltipContext);
 
   const state = mergeProps(
     {
       ref,
-      as: 'div',
+      children: props.children,
+      content: {
+        as: React.Fragment,
+      },
+      id: useId('tooltip-'),
       role: 'tooltip',
       position: 'above',
       align: 'center',
       offset: 4,
-      arrow: { as: 'div', children: '' },
-      visible: !!target,
+      showDelay: 250,
+      hideDelay: 250,
+      visible,
+      shouldRenderTooltip: !useIsSSR(),
     },
     defaultProps && resolveShorthandProps(defaultProps, tooltipShorthandProps),
     resolveShorthandProps(props, tooltipShorthandProps),
   );
 
-  if (state.noArrow) {
-    state.arrow = undefined;
-  }
-
-  React.useImperativeHandle(
-    state.componentRef,
-    () => ({
-      show: t => setTarget(t),
-      hide: () => setTarget(null),
-    }),
-    [],
-  );
-
   const theme = useTheme();
-
-  const { targetRef, containerRef, arrowRef } = usePopper({
-    enabled: state.visible,
+  const popper = usePopper({
+    enabled: visible,
     position: state.position,
     align: state.align,
     offset: [0, state.offset + (state.noArrow ? 0 : arrowHeight)],
     arrowPadding: theme?.global ? 2 * parseInt(tooltipBorderRadius(theme), 10) : 0,
   });
 
-  targetRef.current = target;
-  state.ref = useMergedRefs(state.ref, containerRef);
+  state.ref = useMergedRefs(state.ref, popper.containerRef);
+  state.arrowRef = popper.arrowRef;
+
+  const [setDelayTimeout, clearDelayTimeout] = useTimeout();
+
+  const { targetDocument } = useFluent();
+  useIsomorphicLayoutEffect(() => {
+    if (visible) {
+      const hide = () => {
+        setVisible(false);
+        clearDelayTimeout();
+      };
+
+      // Hide any other tooltip that's currently visible, then register this one
+      context.hideVisibleTooltip?.();
+      context.hideVisibleTooltip = hide;
+
+      // Listen for Escape on the document
+      const onDocumentKeyDown = (ev: KeyboardEvent) => {
+        if (ev.key === 'Escape') {
+          hide();
+        }
+      };
+      targetDocument?.addEventListener('keydown', onDocumentKeyDown);
+
+      return () => {
+        targetDocument?.removeEventListener('keydown', onDocumentKeyDown);
+        if (context.hideVisibleTooltip === hide) {
+          context.hideVisibleTooltip = undefined;
+        }
+      };
+    }
+  }, [clearDelayTimeout, context, targetDocument, visible]);
+
+  const hoveredOrFocused = React.useRef(false);
```

Comment:
I can see why we are not using `useState` here. But, is there any reason to use `useRef` instead of a regular old `let hoveredOrFocused = false`?

Reply:
Using a ref allows me to reference the value from the callbacks cached using `React.useCallback` (which is done so that the callbacks set on the trigger element don't change every render). If I used `let`, the cached callbacks could potentially be referring to different copies of the variable after a re-render.

Reply:
Ahh, I see now, thanks for the explanation, it went over my head.